sql: TestValidationWithProtectedTS is flaky

Saw it on like 5 failures on staging branch. E.g. [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=7196927&buildTypeId=Cockroach_UnitTests_BazelUnitTests).
```
No pass/skip/fail event found for test
=== RUN   TestValidationWithProtectedTS
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/7786b2439f6fbd6878e73b0041b1a786/logTestValidationWithProtectedTS808938841
    test_log_scope.go:79: use -show-logs to present logs inline
panic: test timed out after 59m55s
```
Will skip.

Jira issue: CRDB-20977

--- a/pkg/sql/backfill_protected_timestamp_test.go
+++ b/pkg/sql/backfill_protected_timestamp_test.go
@@ -13,6 +13,8 @@ package sql_test
 import (
 	"context"
 	"regexp"
+	"strings"
+	"sync"
 	"testing"
 	"time"
 
@@ -30,6 +32,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -56,7 +59,6 @@ func getFirstStoreReplica(
 
 func TestValidationWithProtectedTS(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 90879, "flaky test")
 	defer log.Scope(t).Close(t)
 	skip.UnderStress(t, "test takes too long")
 	skip.UnderRace(t, "test takes too long")
@@ -64,6 +66,7 @@ func TestValidationWithProtectedTS(t *testing.T) {
 	ctx := context.Background()
 	indexValidationQueryWait := make(chan struct{})
 	indexValidationQueryResume := make(chan struct{})
+	validationQuerySeen := sync.Once{}
 
 	indexScanQuery := regexp.MustCompile(`SELECT count\(1\) FROM \[\d+ AS t\]@\[2\]`)
 	settings := cluster.MakeTestingClusterSettings()
@@ -77,8 +80,11 @@ func TestValidationWithProtectedTS(t *testing.T) {
 					SQLExecutor: &sql.ExecutorTestingKnobs{
 						BeforeExecute: func(ctx context.Context, sql string, descriptors *descs.Collection) {
 							if indexScanQuery.MatchString(sql) {
-								indexValidationQueryWait <- struct{}{}
-								<-indexValidationQueryResume
+								validationQuerySeen.Do(func() {
+									indexValidationQueryWait <- struct{}{}
+									<-indexValidationQueryResume
+								})
+
 							}
 						},
 					},
@@ -113,7 +119,7 @@ func TestValidationWithProtectedTS(t *testing.T) {
 		"SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval ='10ms'",
 		"ALTER DATABASE defaultdb CONFIGURE ZONE USING gc.ttlseconds = 1",
 		"CREATE TABLE t(n int)",
-		"ALTER TABLE t CONFIGURE ZONE USING range_min_bytes = 0, range_max_bytes = 65536, gc.ttlseconds = 1",
+		"ALTER TABLE t CONFIGURE ZONE USING range_min_bytes = 0, range_max_bytes = 67108864, gc.ttlseconds = 1",
 		"INSERT INTO t(n) SELECT * FROM generate_series(1, 250000)",
 	} {
 		_, err := tc.ServerConn(0).Exec(sql)
@@ -131,45 +137,74 @@ func TestValidationWithProtectedTS(t *testing.T) {
 	tableID := getTableID()
 	tableKey := keys.SystemSQLCodec.TablePrefix(tableID)
 
-	go func() {
+	grp := ctxgroup.WithContext(ctx)
+	grp.Go(func() error {
 		<-indexValidationQueryWait
-		getTableRangeIDs := func(t *testing.T) []int64 {
+		getTableRangeIDs := func(t *testing.T) ([]int64, error) {
 			t.Helper()
 			rows, err := dbConn2.QueryContext(ctx, "WITH r AS (SHOW RANGES FROM TABLE t) SELECT range_id FROM r ORDER BY start_key")
-			require.NoError(t, err, "failed to query ranges")
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to query ranges")
+			}
 			var rangeIDs []int64
 			for rows.Next() {
 				var rangeID int64
-				require.NoError(t, rows.Scan(&rangeID), "failed to read row with range id")
+				if err := rows.Scan(&rangeID); err != nil {
+					return nil, errors.Wrapf(err, "failed to read row with range id")
+				}
 				rangeIDs = append(rangeIDs, rangeID)
 			}
-			require.NoError(t, rows.Close())
-			return rangeIDs
+			if err := rows.Close(); err != nil {
+				return nil, err
+			}
+			return rangeIDs, nil
 		}
-		ranges := getTableRangeIDs(t)
-		_, err := dbConn2.ExecContext(ctx, "BEGIN")
-		require.NoError(t, err)
-		_, err = dbConn2.ExecContext(ctx, "SET sql_safe_updates=off")
-		require.NoError(t, err)
-		_, err = dbConn2.ExecContext(ctx, "DELETE FROM t;")
-		require.NoError(t, err)
-		_, err = dbConn2.ExecContext(ctx, "INSERT INTO t VALUES('9999999')")
-		require.NoError(t, err)
-		_, err = dbConn2.ExecContext(ctx, "COMMIT")
-		require.NoError(t, err)
+		ranges, err := getTableRangeIDs(t)
+		if err != nil {
+			return err
+		}
+		const retryTxnErrorSubstring = "restart transaction"
+		for {
+			if _, err := dbConn2.ExecContext(ctx, "BEGIN"); err != nil {
+				return err
+			}
+			if _, err := dbConn2.ExecContext(ctx, "SET sql_safe_updates=off"); err != nil {
+				return err
+			}
+			if _, err := dbConn2.ExecContext(ctx, "DELETE FROM t;"); err != nil {
+				return err
+			}
+			if _, err := dbConn2.ExecContext(ctx, "INSERT INTO t VALUES('9999999')"); err != nil {
+				return err
+			}
+			_, err = dbConn2.ExecContext(ctx, "COMMIT")
+			if err != nil {
+				if strings.Contains(err.Error(), retryTxnErrorSubstring) {
+					err = nil
+					continue
+				}
+				return err
+			}
+			break
+		}
 		refreshTo(t, tableKey, tc.Server(0).Clock().Now())
 		refreshPTSCacheTo(t, tc.Server(0).Clock().Now())
 		for _, id := range ranges {
-			_, err := dbConn2.ExecContext(ctx, `SELECT crdb_internal.kv_enqueue_replica($1, 'mvccGC', true)`, id)
-			require.NoError(t, err)
+			if _, err := dbConn2.ExecContext(ctx, `SELECT crdb_internal.kv_enqueue_replica($1, 'mvccGC', true)`, id); err != nil {
+				return err
+			}
 		}
 		indexValidationQueryResume <- struct{}{}
+		return nil
+	})
+	grp.Go(func() error {
+		_, err := db.ExecContext(ctx, `CREATE INDEX foo ON t (n)`)
+		return err
+	})
 
-	}()
-	if _, err := db.ExecContext(ctx, `CREATE INDEX foo ON t (n)`); err != nil {
-		t.Fatal(err)
-	}
-	// Validate the rows were removed due to the drop..
+	err := grp.Wait()
+	require.NoError(t, err)
+	// Validate the rows were removed due to the drop
 	res := r.QueryStr(t, `SELECT n FROM t@foo`)
 	if len(res) != 1 {
 		t.Errorf("expected %d entries, got %d", 1, len(res))

--- a/pkg/sql/index_backfiller.go
+++ b/pkg/sql/index_backfiller.go
@@ -13,6 +13,7 @@ package sql
 import (
 	"context"
 
+	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
@@ -27,6 +28,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/admission/admissionpb"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
 )
 
 // IndexBackfillPlanner holds dependencies for an index backfiller
@@ -72,8 +74,22 @@ func (ib *IndexBackfillPlanner) BackfillIndexes(
 	ctx context.Context,
 	progress scexec.BackfillProgress,
 	tracker scexec.BackfillerProgressWriter,
+	job *jobs.Job,
 	descriptor catalog.TableDescriptor,
-) error {
+) (err error) {
+	// Potentially install a protected timestamp before the GC interval is hit,
+	// which can help avoid transaction retry errors, with shorter GC intervals.
+	protectedTimestampCleaner := ib.execCfg.ProtectedTimestampManager.TryToProtectBeforeGC(ctx,
+		job,
+		descriptor,
+		progress.MinimumWriteTimestamp)
+	defer func() {
+		cleanupError := protectedTimestampCleaner(ctx)
+		if cleanupError != nil {
+			err = errors.CombineErrors(cleanupError, err)
+		}
+	}()
+
 	var completed = struct {
 		syncutil.Mutex
 		g roachpb.SpanGroup

--- a/pkg/sql/mvcc_backfiller.go
+++ b/pkg/sql/mvcc_backfiller.go
@@ -46,10 +46,11 @@ func NewIndexBackfillerMergePlanner(execCfg *ExecutorConfig) *IndexBackfillerMer
 // MergeIndexes is part of the scexec.Merger interface.
 func (im *IndexBackfillerMergePlanner) MergeIndexes(
 	ctx context.Context,
+	job *jobs.Job,
 	progress scexec.MergeProgress,
 	tracker scexec.BackfillerProgressWriter,
 	descriptor catalog.TableDescriptor,
-) error {
+) (err error) {
 	spansToDo := make([][]roachpb.Span, len(progress.SourceIndexIDs))
 	if len(progress.CompletedSpans) != len(progress.SourceIndexIDs) {
 		return errors.AssertionFailedf("invalid MergeProgress, CompletedSpans should " +
@@ -96,6 +97,15 @@ func (im *IndexBackfillerMergePlanner) MergeIndexes(
 		)
 		return tracker.SetMergeProgress(ctx, progress)
 	}
+	mergeTimeStamp := getMergeTimestamp(im.execCfg.Clock)
+	protectedTimestampCleaner := im.execCfg.ProtectedTimestampManager.TryToProtectBeforeGC(ctx, job, descriptor, mergeTimeStamp)
+	defer func() {
+		cleanupError := protectedTimestampCleaner(ctx)
+		if cleanupError != nil {
+			err = errors.CombineErrors(cleanupError, err)
+		}
+	}()
+
 	run, err := im.plan(
 		ctx,
 		descriptor,
@@ -103,7 +113,7 @@ func (im *IndexBackfillerMergePlanner) MergeIndexes(
 		progress.DestIndexIDs,
 		progress.SourceIndexIDs,
 		updateFunc,
-		getMergeTimestamp(im.execCfg.Clock),
+		mergeTimeStamp,
 	)
 	if err != nil {
 		return err

--- a/pkg/sql/schemachanger/scdeps/sctestdeps/backfill.go
+++ b/pkg/sql/schemachanger/scdeps/sctestdeps/backfill.go
@@ -13,6 +13,7 @@ package sctestdeps
 import (
 	"context"
 
+	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scexec"
@@ -85,6 +86,7 @@ func (s *testBackfiller) BackfillIndexes(
 	_ context.Context,
 	progress scexec.BackfillProgress,
 	_ scexec.BackfillerProgressWriter,
+	_ *jobs.Job,
 	tbl catalog.TableDescriptor,
 ) error {
 	s.s.LogSideEffectf(
@@ -106,6 +108,7 @@ var _ scexec.Merger = (*testBackfiller)(nil)
 // MergeIndexes implements the scexec.Merger interface.
 func (s *testBackfiller) MergeIndexes(
 	_ context.Context,
+	job *jobs.Job,
 	progress scexec.MergeProgress,
 	_ scexec.BackfillerProgressWriter,
 	tbl catalog.TableDescriptor,

--- a/pkg/sql/schemachanger/scexec/dependencies.go
+++ b/pkg/sql/schemachanger/scexec/dependencies.go
@@ -173,6 +173,7 @@ type Backfiller interface {
 		context.Context,
 		BackfillProgress,
 		BackfillerProgressWriter,
+		*jobs.Job,
 		catalog.TableDescriptor,
 	) error
 }
@@ -184,12 +185,7 @@ type Merger interface {
 
 	// MergeIndexes will merge the specified indexes in the table, from each
 	// temporary index into each adding index.
-	MergeIndexes(
-		context.Context,
-		MergeProgress,
-		BackfillerProgressWriter,
-		catalog.TableDescriptor,
-	) error
+	MergeIndexes(context.Context, *jobs.Job, MergeProgress, BackfillerProgressWriter, catalog.TableDescriptor) error
 }
 
 // Validator provides interfaces that allow indexes and check constraints to be validated.

--- a/pkg/sql/schemachanger/scexec/exec_backfill.go
+++ b/pkg/sql/schemachanger/scexec/exec_backfill.go
@@ -14,6 +14,7 @@ import (
 	"context"
 	"sort"
 
+	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
@@ -43,7 +44,7 @@ func executeBackfillOps(ctx context.Context, deps Dependencies, execute []scop.O
 	if err != nil {
 		return err
 	}
-	return runBackfiller(ctx, deps, tracker, backfillProgresses, mergeProgresses, tables)
+	return runBackfiller(ctx, deps, tracker, backfillProgresses, mergeProgresses, deps.TransactionalJobRegistry().CurrentJob(), tables)
 }
 
 func getTableDescriptorsForBackfillsAndMerges(
@@ -309,6 +310,7 @@ func runBackfiller(
 	tracker BackfillerTracker,
 	backfillProgresses []BackfillProgress,
 	mergeProgresses []MergeProgress,
+	job *jobs.Job,
 	tables map[descpb.ID]catalog.TableDescriptor,
 ) error {
 	if deps.GetTestingKnobs() != nil &&
@@ -324,10 +326,10 @@ func runBackfiller(
 	im := deps.IndexMerger()
 	const op = "run backfills and merges"
 	bf := func(ctx context.Context, p *BackfillProgress) error {
-		return runBackfill(ctx, deps.IndexSpanSplitter(), ib, *p, tracker, tables[p.TableID])
+		return runBackfill(ctx, deps.IndexSpanSplitter(), ib, *p, tracker, job, tables[p.TableID])
 	}
 	mf := func(ctx context.Context, p *MergeProgress) error {
-		return im.MergeIndexes(ctx, *p, tracker, tables[p.TableID])
+		return im.MergeIndexes(ctx, job, *p, tracker, tables[p.TableID])
 	}
 	if err := forEachProgressConcurrently(ctx, op, backfillProgresses, mergeProgresses, bf, mf); err != nil {
 		pgCode := pgerror.GetPGCode(err)
@@ -359,7 +361,8 @@ func runBackfill(
 	backfiller Backfiller,
 	progress BackfillProgress,
 	tracker BackfillerProgressWriter,
+	job *jobs.Job,
 	table catalog.TableDescriptor,
 ) error {
-	return backfiller.BackfillIndexes(ctx, progress, tracker, table)
+	return backfiller.BackfillIndexes(ctx, progress, tracker, job, table)
 }

--- a/pkg/sql/schemachanger/scexec/exec_backfill_test.go
+++ b/pkg/sql/schemachanger/scexec/exec_backfill_test.go
@@ -160,7 +160,7 @@ func TestExecBackfiller(t *testing.T) {
 				FlushCheckpoint(gomock.Any()).
 				After(setProgress)
 			backfillCall := bf.EXPECT().
-				BackfillIndexes(gomock.Any(), scanned, bt, desc).
+				BackfillIndexes(gomock.Any(), scanned, bt, deps.TransactionalJobRegistry().CurrentJob(), desc).
 				After(flushAfterScan)
 			bt.EXPECT().
 				FlushCheckpoint(gomock.Any()).
@@ -245,10 +245,10 @@ func TestExecBackfiller(t *testing.T) {
 					FlushCheckpoint(gomock.Any()).
 					After(setProgress)
 				backfillBarCall := bf.EXPECT().
-					BackfillIndexes(gomock.Any(), scannedBar, bt, bar).
+					BackfillIndexes(gomock.Any(), scannedBar, bt, deps.TransactionalJobRegistry().CurrentJob(), bar).
 					After(flushAfterScan)
 				backfillFooCall := bf.EXPECT().
-					BackfillIndexes(gomock.Any(), progressFoo, bt, foo).
+					BackfillIndexes(gomock.Any(), progressFoo, bt, deps.TransactionalJobRegistry().CurrentJob(), foo).
 					After(flushAfterScan)
 				bt.EXPECT().
 					FlushCheckpoint(gomock.Any()).
@@ -306,7 +306,7 @@ func TestExecBackfiller(t *testing.T) {
 				GetMergeProgress(gomock.Any(), merge).
 				Return(progress, nil)
 			mergeCall := m.EXPECT().
-				MergeIndexes(gomock.Any(), progress, bt, desc).
+				MergeIndexes(gomock.Any(), deps.CurrentJob(), progress, bt, desc).
 				After(getProgress)
 			bt.EXPECT().
 				FlushCheckpoint(gomock.Any()).

--- a/pkg/sql/schemachanger/scexec/executor_external_test.go
+++ b/pkg/sql/schemachanger/scexec/executor_external_test.go
@@ -412,6 +412,7 @@ func (n noopBackfiller) BackfillIndexes(
 	ctx context.Context,
 	progress scexec.BackfillProgress,
 	writer scexec.BackfillerProgressWriter,
+	job *jobs.Job,
 	descriptor catalog.TableDescriptor,
 ) error {
 	return nil
@@ -441,6 +442,7 @@ var _ scexec.Merger = (*noopMerger)(nil)
 
 func (n noopMerger) MergeIndexes(
 	ctx context.Context,
+	job *jobs.Job,
 	progress scexec.MergeProgress,
 	writer scexec.BackfillerProgressWriter,
 	descriptor catalog.TableDescriptor,

--- a/pkg/sql/schemachanger/scexec/mocks_generated_test.go
+++ b/pkg/sql/schemachanger/scexec/mocks_generated_test.go
@@ -8,6 +8,7 @@ import (
 	context "context"
 	reflect "reflect"
 
+	jobs "github.com/cockroachdb/cockroach/pkg/jobs"
 	username "github.com/cockroachdb/cockroach/pkg/security/username"
 	cluster "github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	catalog "github.com/cockroachdb/cockroach/pkg/sql/catalog"
@@ -503,17 +504,17 @@ func (m *MockBackfiller) EXPECT() *MockBackfillerMockRecorder {
 }
 
 // BackfillIndexes mocks base method.
-func (m *MockBackfiller) BackfillIndexes(arg0 context.Context, arg1 scexec.BackfillProgress, arg2 scexec.BackfillerProgressWriter, arg3 catalog.TableDescriptor) error {
+func (m *MockBackfiller) BackfillIndexes(arg0 context.Context, arg1 scexec.BackfillProgress, arg2 scexec.BackfillerProgressWriter, arg3 *jobs.Job, arg4 catalog.TableDescriptor) error {
 	m.ctrl.T.Helper()
-	ret := m.ctrl.Call(m, "BackfillIndexes", arg0, arg1, arg2, arg3)
+	ret := m.ctrl.Call(m, "BackfillIndexes", arg0, arg1, arg2, arg3, arg4)
 	ret0, _ := ret[0].(error)
 	return ret0
 }
 
 // BackfillIndexes indicates an expected call of BackfillIndexes.
-func (mr *MockBackfillerMockRecorder) BackfillIndexes(arg0, arg1, arg2, arg3 interface{}) *gomock.Call {
+func (mr *MockBackfillerMockRecorder) BackfillIndexes(arg0, arg1, arg2, arg3, arg4 interface{}) *gomock.Call {
 	mr.mock.ctrl.T.Helper()
-	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "BackfillIndexes", reflect.TypeOf((*MockBackfiller)(nil).BackfillIndexes), arg0, arg1, arg2, arg3)
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "BackfillIndexes", reflect.TypeOf((*MockBackfiller)(nil).BackfillIndexes), arg0, arg1, arg2, arg3, arg4)
 }
 
 // MaybePrepareDestIndexesForBackfill mocks base method.
@@ -555,17 +556,17 @@ func (m *MockMerger) EXPECT() *MockMergerMockRecorder {
 }
 
 // MergeIndexes mocks base method.
-func (m *MockMerger) MergeIndexes(arg0 context.Context, arg1 scexec.MergeProgress, arg2 scexec.BackfillerProgressWriter, arg3 catalog.TableDescriptor) error {
+func (m *MockMerger) MergeIndexes(arg0 context.Context, arg1 *jobs.Job, arg2 scexec.MergeProgress, arg3 scexec.BackfillerProgressWriter, arg4 catalog.TableDescriptor) error {
 	m.ctrl.T.Helper()
-	ret := m.ctrl.Call(m, "MergeIndexes", arg0, arg1, arg2, arg3)
+	ret := m.ctrl.Call(m, "MergeIndexes", arg0, arg1, arg2, arg3, arg4)
 	ret0, _ := ret[0].(error)
 	return ret0
 }
 
 // MergeIndexes indicates an expected call of MergeIndexes.
-func (mr *MockMergerMockRecorder) MergeIndexes(arg0, arg1, arg2, arg3 interface{}) *gomock.Call {
+func (mr *MockMergerMockRecorder) MergeIndexes(arg0, arg1, arg2, arg3, arg4 interface{}) *gomock.Call {
 	mr.mock.ctrl.T.Helper()
-	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "MergeIndexes", reflect.TypeOf((*MockMerger)(nil).MergeIndexes), arg0, arg1, arg2, arg3)
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "MergeIndexes", reflect.TypeOf((*MockMerger)(nil).MergeIndexes), arg0, arg1, arg2, arg3, arg4)
 }
 
 // MockBackfillerTracker is a mock of BackfillerTracker interface.

